src/ray/raylet/raylet_client.cc

```diff
@@ -198,11 +96,11 @@ raylet::RayletClient::RayletClient(
   fbb.Finish(message);
   // Register the process ID with the raylet.
   // NOTE(swang): If raylet exits and we are registered as a worker, we will get killed.
-  std::unique_ptr<uint8_t[]> reply;
+  std::vector<uint8_t> reply;
```

Comment:
There are 4 points for me to use `std::vector` instead.

1. I find it is not convenient to use `std::unique_ptr<uint8_t[]>` in general cases with `boost::asio`. `boost::asio` has native support for `std::vector` as a buffer. To use `std::unique_ptr<uint8_t[]>` with `boost::asio`, we need to use `get()` to extract the underlying raw pointer, and pass the length of the buffer as an independent parameter. In all other places, we have replaced `std::unique_ptr<uint8_t[]>` with `std::vector`.

2. For our general message processing, we need a buffer with a length for the purpose of various data parsing (flatbuffer, msgpack, protobuf, etc). For me `std::vector` is a better abstraction with some boundary guarantee.

3. The performance is not a problem, because we need to allocating memory for `std::unique_ptr<uint8_t[]>` dynamically too, and we use `reserve` for `std::vector`.

4. Google C++ style prefers pointer type for `output arguments`. This creates ugly types like `std::unique_ptr<uint8_t[]>*`. I really don't like it. Why not just use managed buffers like `std::vector` to control the lifetime? C++ should ensure zero-cost abstraction.

Reply:
1. [Asio's primary support](https://www.boost.org/doc/libs/1_73_0/doc/html/boost_asio/overview/core/buffers.html) is for e.g. [`mutable_buffer`](https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/buffer.html) which is a pointer and a size, not vector.

2. I don't think we need `size()` though, do we? We weren't using it so far (given `std::unique_ptr` sufficed for our usage), and even after the change, I see
    ```
    flatbuffers::GetRoot<protocol::RegisterClientReply>(reply.data());
    ```
    which still ignores the size. I might be missing something though.

3. Vector has additional initialization overhead (I'm not referring to allocation) compared to a plain dynamic array.

4. I agree `std::unique_ptr<uint8_t[]> *` isn't pretty, but `std::vector<uint8_t> *` is the same thing with size/capacity added. The double-nullity is nice to avoid, but you can avoid that with a reference rather than a pointer. In fact, the Google style guide (which I don't really defer to personally) [says the same thing](https://google.github.io/styleguide/cppguide.html): _"required (non-nullable) output and input/output parameters should usually be references"_.

Personally my biggest concern though is that `vector` is semantically meant to be a random-access container and not a buffer of raw memory, so it ends up being somewhat misleading IMO. But that said, I don't feel too strongly about it, so if you disagree despite these feel free to keep it.

Reply:
The size is used everywhere in Plasma Store, which is very painful if we only keep `std::unique_ptr<uint8_t[]> *`. You can see the `MessageHandler` in ClientConnection has been using `std::vector`.

https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/src/ray/object_manager/plasma/store.cc#L894-L916

We need to use size here:

https://github.com/ray-project/ray/blob/8c985dc79764408be7004f327a6c0f5a625b2271/src/ray/object_manager/plasma/protocol.h#L36-L40